ibis/backends/base/sqlglot/compiler.py

```diff
@@ -192,22 +206,168 @@ class SQLGlotCompiler(abc.ABC):
+        ops.StringLength: "length",
+        ops.StringReplace: "replace",
+        ops.StringSplit: "split",
+        ops.StringToTimestamp: "str_to_time",
+        ops.Tan: "tan",
+        ops.Translate: "translate",
+        ops.Unnest: "explode",
+        ops.Uppercase: "upper",
+    }
+
+    BINARY_INFIX_OPS = (
+        # Binary operations
+        ops.Add,
+        ops.Subtract,
+        ops.Multiply,
+        ops.Divide,
+        ops.Modulus,
+        ops.Power,
+        # Comparisons
+        ops.GreaterEqual,
+        ops.Greater,
+        ops.LessEqual,
+        ops.Less,
+        ops.Equals,
+        ops.NotEquals,
+        # Boolean comparisons
+        ops.And,
+        ops.Or,
+        ops.Xor,
+        # Bitwise business
+        ops.BitwiseLeftShift,
+        ops.BitwiseRightShift,
+        ops.BitwiseAnd,
+        ops.BitwiseOr,
+        ops.BitwiseXor,
+        # Time arithmetic
+        ops.DateAdd,
+        ops.DateSub,
+        ops.DateDiff,
+        ops.TimestampAdd,
+        ops.TimestampSub,
+        ops.TimestampDiff,
+        # Interval Marginalia
+        ops.IntervalAdd,
+        ops.IntervalMultiply,
+        ops.IntervalSubtract,
+    )
+
     def __init__(self) -> None:
         self.agg = AggGen(aggfunc=self._aggregate)
         self.f = FuncGen()
         self.v = VarGen()
 
     def __init_subclass__(cls, **kwargs):
-        for leaf in ALL_OPERATIONS:
-            if not hasattr(cls, f"visit_{leaf.__name__}"):
-                setattr(cls, f"visit_{leaf.__name__}", cls.visit_Undefined)
+        super().__init_subclass__(**kwargs)
+
+        def methodname(op: type) -> str:
+            assert isinstance(type(op), type), type(op)
+            return f"visit_{op.__name__}"
+
+        def make_impl(op, target_name):
+            assert isinstance(type(op), type), type(op)
 
-        for leaf in cls.UNSUPPORTED_OPERATIONS:
+            if issubclass(op, ops.Reduction):
+
+                def impl(self, _, *, _name: str = target_name, where, **kw):
+                    return self.agg[_name](*kw.values(), where=where)
+
+            else:
+
+                def impl(self, _, *, _name: str = target_name, **kw):
+                    return self.f[_name](*kw.values())
+
+            return impl
+
+        # unconditionally raise an exception for unsupported operations
+        for op in cls.UNSUPPORTED_OPERATIONS:
             # change to visit_Unsupported in a follow up
             # TODO: handle geoespatial ops as a separate case?
-            setattr(cls, f"visit_{leaf.__name__}", cls.visit_Undefined)
+            setattr(cls, methodname(op), cls.visit_Undefined)
 
-        super().__init_subclass__(**kwargs)
+        # override existing base class implementations
+        for op, target_name in cls.SIMPLE_OPS.items():
+            setattr(cls, methodname(op), make_impl(op, target_name))
+
+        # add simple ops that are not already implemented
+        for op, target_name in SQLGlotCompiler.SIMPLE_OPS.items():
```

Comment:
Why is this required?

Reply:
So that all child classes have these operations.

Reply:
Ohh, I thought these are already available on the class.